File: kotlin/0242-e-valid-anagram.kt
// https://leetcode.com/problems/valid-anagram
class Solution {
    fun isAnagram(s: String, t: String): Boolean {
        if (s.length != t.length) return false
        val sMap = HashMap<Char, Int>()
        val tMap = HashMap<Char, Int>()

        for (i in 0 until s.length) {
            sMap[s[i]] = sMap.getOrDefault(s[i], 0) + 1
            tMap[t[i]] = tMap.getOrDefault(t[i], 0) + 1
        }

        for ((key, value) in sMap) {
            if (tMap.getOrDefault(key, 0) != value) return false
        }

        return true
    }
}

File: python/0242-e-valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t): return False
        sMap = {}
        tMap = {}

        for i, c in enumerate(s):
            sMap[c] = sMap.get(c, 0) + 1
            tMap[t[i]] = tMap.get(t[i], 0) + 1

        for key, val in sMap.items():
            if tMap.get(key, 0) != val: return False

        return True

if __name__ == '__main__':
    print(Solution().isAnagram("anagram","gramana"))
    print(Solution().isAnagram("car","rat"))
    print(Solution().isAnagram("c","car"))